plugins/inputs/loggregator_rlp/rlp.go

```diff
@@ -0,0 +1,178 @@
+package loggregator_rlp
+
+import (
+	"context"
+	"github.com/influxdata/telegraf/internal"
+	"log"
+	"os"
+	"time"
+
+	"github.com/influxdata/telegraf/internal/tls"
+
+	"code.cloudfoundry.org/go-loggregator"
+	"code.cloudfoundry.org/go-loggregator/rpc/loggregator_v2"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type LoggregatorRLPInput struct {
+	TlsCommonName           string            `toml:"tls_common_name"`
+	RlpAddress              string            `toml:"rlp_address"`
+	InternalMetricsInterval internal.Duration `toml:"internal_metrics_interval"`
+	DiodeBufferSize         int               `toml:"diode_buffer_size"`
+	EnvelopeTypes           []string          `toml:"envelope_types"`
+	SourceIdFilters         []string          `toml:"source_id_filters"`
+	stopRlpConsumer         context.CancelFunc
+	envelopeWriter          *EnvelopeWriter
+
+	tls.ClientConfig
+}
+
+func NewLoggregatorRLP() *LoggregatorRLPInput {
+	return &LoggregatorRLPInput{
+		InternalMetricsInterval: internal.Duration{
+			Duration: 30 * time.Second,
+		},
+	}
+}
+
+func (_ *LoggregatorRLPInput) Description() string {
+	return "Streams metrics from Loggregator's RLP Endpoint"
+}
+
+func (_ *LoggregatorRLPInput) SampleConfig() string {
+	return `
+  ## A string path to the tls ca certificate
+  tls_ca = "/path/to/tls_ca_cert.pem"
+
+  ## A string path to the tls server certificate
+  tls_cert = "/path/to/tls_cert.pem"
+
+  ## A string path to the tls server private key
+  tls_key = "/path/to/tls_cert.key"
+
+  ## Boolean value indicating whether or not to skip SSL verification
+  insecure_skip_verify = false
+  
+  ## A string server name that the certificate is valid for
+  tls_common_name = "foo"
+  
+  ## A string address of the RLP server to get logs from
+  rlp_address = "bar"
+
+  ## A string duration for how frequently to report internal metrics
+  internal_metrics_interval = "30s"
+
+  ## Size of diode buffer, this is the limit of how many envelopes will be held in memory before dropping new envelopes
+  diode_buffer_size = 100000
+
+  ## Envelope type, an array of one or all of ['counter', 'gauge', 'timer']
+  envelope_types = ['counter', 'gauge']
+
+  ## Source Id filters allow the filtering of metrics out of the RLP from only the specified source ids. Should be an array of strings.
+  source_id_filters = ['router']
+`
+}
+
+func (l *LoggregatorRLPInput) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (l *LoggregatorRLPInput) Start(acc telegraf.Accumulator) error {
+	envelopeWriter := NewEnvelopeWriter(acc, l.InternalMetricsInterval.Duration)
```

Comment:
I don't see the purpose of this, just call a function in the consumer goroutine on line 103 that will add the data to the accumulator.  This will dramatically simplify this plugin and eliminate the need for the internal metrics and ring buffer.  The ring buffer duplicates functionality that Telegraf gives you for free and introduces another buffer and with it the potential for dropped metrics.